stdlib/mktemp.c

```diff
@@ -28,43 +28,58 @@ static char pfcs[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz012345
 
 
 /* Generates a unique filename */
-static int mktemp(char *templt, int suffixlen)
+static int __mktemp(char *templt, int suffixlen)
 {
-	int i, len, rand, fd;
-	char *tail;
-	oid_t oid;
-
-	if ((templt == NULL) || (suffixlen < 0))
+	if ((templt == NULL) || (suffixlen < 0)) {
 		return -EINVAL;
+	}
 
-	len = strlen(templt);
-	tail = templt + len - suffixlen - 6;
+	int len = strlen(templt);
+	char *tail = templt + len - suffixlen - 6;
 
-	if ((len < suffixlen + 6) || strncmp(tail, "XXXXXX", 6))
+	if ((len < (suffixlen + 6)) || (strncmp(tail, "XXXXXX", 6) != 0)) {
 		return -EINVAL;
+	}
 
-	if ((fd = open("/dev/urandom", O_RDONLY)) < 0)
+	int fd = open("/dev/urandom", O_RDONLY);
+	if (fd < 0) {
 		return fd;
-
-	for (i = 0; i < 6; i++) {
-		read(fd, &rand, sizeof(rand));
-		tail[i] = pfcs[rand % sizeof(pfcs)];
 	}
 
+	int rand[6];
+	ssize_t left = sizeof(rand);
+	do {
+		ssize_t err;
+		do {
+			err = read(fd, rand + sizeof(rand) - left, left);
+		} while ((err < 0) && (errno == -EINTR));
+
+		if (err <= 0) {
+			close(fd);
+			return -EIO;
+		}
+		left -= err;
+	} while (left != 0);
 	close(fd);
 
-	if (lookup(templt, &oid, NULL) == EOK)
+	for (int i = 0; i < 6; ++i) {
+		tail[i] = pfcs[rand[i] % (sizeof(pfcs) - 1)];
+	}
+
+	oid_t oid;
+	if (lookup(templt, &oid, NULL) == 0) {
 		return -EEXIST;
+	}
 
-	return EOK;
+	return 0;
 }
 
 
 int mkostemps(char *templt, int suffixlen, int flags)
 {
 	int fd;
 
-	if (mktemp(templt, suffixlen) < 0)
+	if (__mktemp(templt, suffixlen) < 0)
```

Comment:
set errno and return -1? (warn: in one use case there is `return fd`, so maybe only if != -1?)